python/ql/lib/semmle/python/security/dataflow/UnsafeShellCommandConstructionQuery.qll

```diff
@@ -28,6 +28,16 @@ module UnsafeShellCommandConstructionConfig implements DataFlow::ConfigSig {
 
   // override to require the path doesn't have unmatched return steps
   DataFlow::FlowFeature getAFeature() { result instanceof DataFlow::FeatureHasSourceCallContext }
+
+  predicate observeDiffInformedIncrementalMode() { any() }
+
+  Location getASelectedSinkLocation(DataFlow::Node sink) {
+    result = sink.(Sink).getLocation()
+    or
+    result = sink.(Sink).getStringConstruction().getLocation()
+    or
+    result = sink.(Sink).getCommandExecution().getLocation()
```

Comment:
Optional thought: This pattern makes me think we should allow the custom `Sink` classes to expose a `getASelectedLocation` predicate and expand it through overrides, then just call `result = sink.(Sink).getASelectedLocation()` here. Not necessary to figure out for this PR.

Reply:
The pattern I have in mind is that the `select` clause must only select entities that are returned from a member predicate on the `Source` and `Sink` classes, and then those same predicates must be mentioned in the config. This can be enforced with a QL4QL query without needing too much static analysis.

The problem with putting a `getASelectedLocation` member on `Sink` is that it must still be implemented far away from the `select` clause. It doesn't avoid the strong coupling between two things that are far apart.